rewrite-core/src/main/java/org/openrewrite/GitRemote.java

```diff
@@ -46,6 +47,38 @@ public enum Service {
         Unknown
     }
 
+    public URI toUri(boolean ssh) {
+        return buildRemoteUrl(service, ssh, origin, path);
+    }
+
+    public static URI buildRemoteUrl(Service service, boolean ssh, String origin, String path) {
+        StringBuilder url = new StringBuilder();
+        path = path.replaceFirst("^/", "").replaceFirst("/$", "");
+        origin = origin.replaceFirst("/$", "");
+        if (service == Service.Bitbucket && !ssh) {
+            path = "scm/" + path;
+        }
+        if (service == Service.AzureDevOps) {
+            if (ssh) {
+                origin = "ssh." + origin;
+            } else {
+                path = path.replaceFirst("([^/]+)/([^/]+)/(.*)", "$1/$2/_git/$3");
+            }
+        } else if (service == Service.GitLab || ssh) {
+            path += ".git";
+        }
+        if (origin.startsWith("https://") || origin.startsWith("http://") || origin.startsWith("ssh://")) {
+            origin = origin.substring(origin.indexOf("://") + 3);
+        }
+        if (ssh) {
+            url.append("ssh://git@");
```

Comment:
Can you add ADO specific logic here, similar to 
```
         if (ssh) {
                if (scmType == RepositoryScmType.AzureDevOpsRepository) {
                    builder.append(":v3");
                } else if (scmType == RepositoryScmType.BitbucketRepository) {
                    int slashIndex = builder.indexOf("/", originIndex);
                    if (slashIndex != -1) {
                        builder.insert(slashIndex, ":7999");
                    } else {
                        builder.append(":7999");
                    }
                }
            }
```

Reply:
sure, however I do not want to build scp style urls. Even though that is what most SCM's give us I think it is best to build "real" URIs

Reply:
But the URI should be what is expected specific to the SCM no?

Reply:
I did some testing and the SCP syntax works, but for ADO you'll need to add a `v3/` segment. In other words

`ssh://git@ssh.dev.azure.com/moderneinc/moderne/git-test` does not work but `ssh://git@ssh.dev.azure.com/v3/moderneinc/moderne/git-test` does

```
[johnsmith@localhost ~/repos/azure-devops/clone-test2]$ git clone ssh://git@ssh.dev.azure.com/moderneinc/moderne/git-test
Cloning into 'git-test'...
remote: Command git-upload-pack '/moderneinc/moderne/git-test' is not in expected format.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
[johnsmith@localhost ~/repos/azure-devops/clone-test2]$ git clone ssh://git@ssh.dev.azure.com/v3/moderneinc/moderne/git-test
Cloning into 'git-test'...
remote: Azure Repos
remote: Found 31 objects to send. (125 ms)
Receiving objects: 100% (31/31), done.
```

Reply:
Ah, I was using an old diff then....